File: APDB_realdata_n10000/apdb_matlab/APDB_c_xy_restart.m
function [rel_infeas_err,rel_subopt_err,time_period,iter_epoch,oracle]=...
    APDB_c_xy_restart(input,optval,x0,y0,stop_criteria,max_iter,non_mono,K)
A = input{1,1};
Q = input{2,1};
b = input{3,1};
d = input{4,1};
c = input{5,1};
sc = 0;
[n,m] = size(d); % number of constraints
optval_rel = abs(optval)+1;  % used for evaluation
%-------------------------------------%
disp('**********************************************************')
disp('APDB-xy-restarts')
disp('**********************************************************')
epoch=1;
epoch_counter = 0;
%------------ step-size Parmeters------%
eta = 0.7;
tau = 3e-1;
gamma = 3;gamma_0 = 1;
sigma_old = gamma*tau;
c_alpha = 0.25; c_beta = 0.3; delta = 0.4;
%% be careful about alpha and beta
alpha_old = c_alpha /tau;
beta_old = gamma_0 * c_beta/sigma_old; % because gamma_0 = 1

% because we have both sigma_{k-1} and sigma_k in each iterate
tau_old = tau;
mu = sc;

%------ Initialization ----------------%
rel_subopt_err = [];
rel_infeas_err = [];
y = y0;x = x0;
x_cum = zeros(m,1);y_cum = zeros(n,1); sigma_0 = sigma_old; TK = 0;

iter = 0;
orc = 0; % oracle
elapsed_time = 0;
inner = zeros(max_iter,1); % to record the number of line search in each iteration
G_x = A * x + b;
for j=1:n
    G_x = G_x+(Q{j,1}*x+d(j,:)')*y(j);
end
orc = orc+1;
% for j=1:n
%    G_y(j,1) = 0.5*x'*Q{j,1}*x+d(j,:)*x-c(j);
% end
%---------- Main Algorithm ------------%
while iter<max_iter
    tic;
    iter = iter+1;
    G_x_old = G_x;
    while true   % line search
        sigma = gamma*tau;
        theta = sigma_old/sigma;
        alpha = c_alpha /tau;
        beta = gamma_0 * c_beta/sigma; % because gamma_0 = 1
        %% to update first x, then y
        G_x = A*x+b;
        for j=1:n
            G_x = G_x+(x' * Q{j,1} +d(j,:))'*y(j);
        end
        
        xtild = max(min(x-tau*((1+theta)*G_x - theta*G_x_old),10),-10);
        xdiff = xtild-x;
        for j=1:n
            G_y(j,1) = 0.5*xtild'*Q{j,1}*xtild+d(j,:)*xtild-c(j);
        end
        orc = orc+1;
        ytild = max(y+sigma*(G_y),0);
        ydiff = ytild-y;
        % the new function E_k
        aux1 = zeros(m,1);
        aux2 = A * xdiff;
        for j=1:n
            aux1 = aux1 + (xtild' * Q{j,1} + d(j,:))' *ydiff(j);
            aux2 = aux2 + (Q{j,1}*xdiff) * y(j);
        end
        % E = 1/2/alpha * norm(aux1)^2 + 1/2/beta * norm(aux2)^2 + (delta-1)/2/sigma * norm(ydiff)^2  - ((1-delta)/tau - theta * (alpha_old+beta_old))/2 * norm(xdiff)^2;
        E = 1/2/alpha * norm(aux1)^2 + 1/2/beta * norm(aux2)^2 + (delta-1)/2/sigma * norm(ydiff)^2  - ((1-delta)/tau - theta * (alpha_old+beta_old))/2 * norm(xdiff)^2;
        % E = 1/2/alpha * norm(aux1)^2 + 1/2/beta * norm(aux2)^2 - ((1-delta)/tau - theta * (alpha_old+beta_old))/2 * norm(xdiff)^2;
        if E<=0
            break;
        else
            inner(iter)=inner(iter)+1;
            tau = tau * eta;
        end
    end
    if mod(iter,500)==0
        fprintf('Iter: %d, line search: %d\n', iter, inner(iter));
    end
    gamma_old = gamma;
    gamma = gamma*(1+mu*tau);
    % line search
    if non_mono == 0
        tau_new = tau*sqrt(gamma_old/gamma);
    else
        tau_new = min( tau*sqrt(gamma_old/gamma *(1+tau/tau_old)), 1);
    end
    % tau_new = tau*sqrt(gamma_old/gamma);
    % tau_new = min(tau*sqrt(gamma_old/gamma *(1+tau/tau_old)), 2);
    sigma_old = sigma;
    tau_old = tau;
    alpha_old = alpha;
    beta_old = beta;
    tau = tau_new;
    
    y = ytild;
    x = xtild;
    x_cum = x_cum + sigma/sigma_0 * x; y_cum = y_cum + sigma/sigma_0 * y; TK= TK + sigma/sigma_0;
    % x_cum = x_cum + x; y_cum = y_cum + y; TK= TK + 1;
    %% restarts
    if mod(iter,K) == 0
        fprintf('perform fixed-frequency restart \n');
        % xavg = x_cum/TK; yavg = y_cum/TK;
        xavg = x; yavg = y;
        
        % initialize
        x = xavg;y = yavg;
        x_cum = zeros(m,1);y_cum = zeros(n,1);TK = 0;
        
        % eta = 0.7;
        tau = 3e-1;
        % gamma = 1;
        sigma_old = gamma*tau;
        % c_alpha = 0.25; c_beta = 0.3; delta = 0.4;
        %% be careful about alpha and beta
        alpha_old = c_alpha /tau;
        beta_old = gamma_0 * c_beta/sigma_old; % because gamma_0 = 1
        
        % because we have both sigma_{k-1} and sigma_k in each iterate
        tau_old = tau;
        G_x = A * x + b;
        for j=1:n
            G_x = G_x+(Q{j,1}*x+d(j,:)')*y(j);
        end
    end
    elapsed_time = elapsed_time + toc;
    if mod(iter,epoch) == 0
        epoch_counter = epoch_counter+1;
        time_period(epoch_counter,1) = elapsed_time;
        oracle(epoch_counter,1) = orc;
        subopt = 0.5*x'*A'*x+b'*x;
        infeas = 0;
        for l=1:n
            infeas = infeas+pos(0.5*x'*Q{l,:}*x+d(l,:)*x-c(l));
        end
        rel_subopt_err(epoch_counter,1) = abs(subopt-optval)/optval_rel;
        rel_infeas_err(epoch_counter,1) = infeas/n;
        iter_epoch(epoch_counter,1) = iter;
        if max(abs(subopt-optval)/abs(optval_rel), infeas/n)<stop_criteria
            fprintf(...
                'Iteration    Time    Rel. Infeas error   Rel. Subopt error\n');
            fprintf('%d    %9.4f       %9.1e         %9.1e\n',iter,time_period(epoch_counter),...
                rel_infeas_err(epoch_counter),rel_subopt_err(epoch_counter));
            return;
        end
    end
end
fprintf(...
    'Iteration    Time    Rel. Infeas error   Rel. Subopt error\n');
fprintf('%d    %9.4f       %9.1e         %9.1e\n',iter,time_period(epoch_counter),...
    rel_infeas_err(epoch_counter),rel_subopt_err(epoch_counter));
end

File: APDB_random_qcqp/apdb_python/solvers/apdb_c_xy_restart.py
from __future__ import annotations

import time

import numpy as np
import torch

from apdb_python.core import project_box, to_numpy, to_torch_problem
from apdb_python.solvers.common import compute_g_x, compute_g_y, evaluate_metrics, finalize_history


def run_apdb_c_xy_restart(
    input_data: dict,
    optval: float,
    x0: np.ndarray,
    y0: np.ndarray,
    stop_criteria: float,
    max_iter: int,
    non_mono: int,
    k_restart: int,
    device: str = "cpu",
    gamma0: float = 1.0,
    gamma_0: float = 1.0,
    ls_debug: bool = False,
) -> dict:
    prob = to_torch_problem(input_data, device=device)
    n_con, m_vars = prob.d.shape

    eta = 0.7
    tau = 3e-1
    gamma = float(gamma0)
    gamma_0 = float(gamma_0)
    c_alpha = 0.25
    c_beta = 0.3
    delta = 0.4
    sigma_old = gamma * tau
    alpha_old = c_alpha / tau
    beta_old = gamma_0 * c_beta / sigma_old
    tau_old = tau
    mu = 0.0

    x = torch.as_tensor(x0, dtype=torch.float64, device=device).reshape(-1)
    y = torch.as_tensor(y0, dtype=torch.float64, device=device).reshape(-1)

    x_cum = torch.zeros(m_vars, dtype=torch.float64, device=device)
    y_cum = torch.zeros(n_con, dtype=torch.float64, device=device)
    sigma_0 = sigma_old
    tk = 0.0

    g_x = compute_g_x(x, y, prob)
    orc = 1

    iter_idx = 0
    elapsed = 0.0
    inner = np.zeros(max_iter, dtype=np.float64)
    optval_rel = abs(float(optval)) + 1.0

    history = {k: [] for k in ["rel_infeas_err", "rel_subopt_err", "time_period", "iter_epoch", "oracle"]}

    while iter_idx < max_iter:
        t0 = time.perf_counter()
        iter_idx += 1
        g_x_old = g_x

        while True:
            sigma = gamma * tau
            theta = sigma_old / sigma
            alpha = c_alpha / tau
            beta = gamma_0 * c_beta / sigma

            g_x = compute_g_x(x, y, prob)
            x_tild = project_box(x - tau * ((1.0 + theta) * g_x - theta * g_x_old))
            xdiff = x_tild - x

            g_y = compute_g_y(x_tild, prob)
            orc += 1
            y_tild = torch.clamp(y + sigma * g_y, min=0.0)
            ydiff = y_tild - y

            aux1 = torch.zeros(m_vars, dtype=torch.float64, device=device)
            aux2 = prob.A @ xdiff
            for j, q_j in enumerate(prob.Q):
                aux1 = aux1 + (q_j @ x_tild + prob.d[j]) * ydiff[j]
                aux2 = aux2 + (q_j @ xdiff) * y[j]

            e_val = (
                0.5 * aux1.norm().pow(2) / alpha
                + 0.5 * aux2.norm().pow(2) / beta
                + (delta - 1.0) * ydiff.norm().pow(2) / (2.0 * sigma)
                - ((1.0 - delta) / tau - theta * (alpha_old + beta_old)) * xdiff.norm().pow(2) / 2.0
            )

            if float(e_val) <= 0.0:
                break
            inner[iter_idx - 1] += 1
            tau = tau * eta
            if ls_debug:
                print(
                    f"[rAPDB-xy][mode={int(non_mono)}][iter={iter_idx}] "
                    f"ls_step={int(inner[iter_idx - 1])} tau={tau:.6e}"
                )

        gamma_old = gamma
        gamma = gamma * (1.0 + mu * tau)
        if non_mono == 0:
            tau_new = tau * np.sqrt(gamma_old / gamma)
        else:
            tau_new = min(tau * np.sqrt(gamma_old / gamma * (1.0 + tau / tau_old)), 1.0)

        sigma_old = sigma
        tau_old = tau
        alpha_old = alpha
        beta_old = beta
        tau = tau_new

        x = x_tild
        y = y_tild

        w = sigma / sigma_0
        x_cum = x_cum + w * x
        y_cum = y_cum + w * y
        tk += w

        if iter_idx % k_restart == 0:
            x = x
            y = y
            x_cum = torch.zeros_like(x_cum)
            y_cum = torch.zeros_like(y_cum)
            tk = 0.0

            tau = 3e-1
            sigma_old = gamma * tau
            alpha_old = c_alpha / tau
            beta_old = gamma_0 * c_beta / sigma_old
            tau_old = tau
            g_x = compute_g_x(x, y, prob)

        elapsed += time.perf_counter() - t0

        rel_sub, infeas = evaluate_metrics(x, prob, optval, optval_rel)
        history["rel_subopt_err"].append(rel_sub)
        history["rel_infeas_err"].append(infeas)
        history["time_period"].append(elapsed)
        history["oracle"].append(float(orc))
        history["iter_epoch"].append(float(iter_idx))

        if max(rel_sub, infeas) < stop_criteria:
            break

    out = finalize_history(history)
    out["x_final"] = to_numpy(x)
    out["y_final"] = to_numpy(y)
    out["inner_ls"] = inner[:iter_idx]
    return out
